### DISCUSSION TITLE
How to import NPM module published as ES2020 or transpile it at build time ? (Lit SSR)

### DISCUSSION THREAD

**Discussion by PonomareVlad** [vercel/vercel#6194] [Help] [answered]
Salut! 🙂

Environment: pure Node.js 14 application (without typescript or babel).

[Demo repo](https://github.com/PonomareVlad/VerceLitSSR/tree/test) (and [Dirty working solution *](https://github.com/PonomareVlad/VerceLitSSR/tree/master)).

I want to deploy a project based on [Lit](https://github.com/lit/lit) [SSR](https://github.com/lit/lit/tree/main/packages/labs/ssr), but I am faced with an _import_ error, since the packages is published as `{"type": "module"}` and **not transpiled** to use as CJS (`require` instead `import`).

> Error [ERR_REQUIRE_ESM]: Must use import to load ES Module ... require() of ES modules is not supported.
> ... is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
> Instead rename render-with-global-dom-shim.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /var/task/node_modules/@lit-labs/ssr/package.json.

![Снимок экрана 2021-05-07 в 02 12 56](https://user-images.githubusercontent.com/2877584/117366357-10081200-aeda-11eb-87cc-6aa9c6d5f339.png)

> \* In the sources of the [@vercel/node](https://github.com/vercel/vercel/tree/main/packages/node) package, I found a condition that prohibits automatic transpilation of the contents of **node_modules** during the build of a lambda function.
> 
> I tried to work around this limitation by publishing the required modules along with the project.
> 
> Also I removed `{"type": "module"}` from the configuration of the required modules and fixed the import paths between them.

**Comment by styfle** [comment, marked as answer]
Hi @PonomareVlad 

We don't currently support `type=module` or `.mjs` (see #5864) for Serverless Functions but we will soon, thanks!

**Reply by styfle** [reply]
Hi @PonomareVlad

We have experimental support for ESM on the canary channel.

You can try it out now by adding the environment variable `FORCE_RUNTIME_TAG=canary` to your project and deploying again.

**Reply by PonomareVlad** [reply]
Thank you @styfle !

I just tried to deploy the project with the specified ENV, but the execution ends with the same errors 😟

![Снимок экрана 2021-06-24 в 22 11 08](https://user-images.githubusercontent.com/2877584/123305152-3725ac00-d539-11eb-80c4-924128aef979.png)

![Снимок экрана 2021-06-24 в 21 41 09](https://user-images.githubusercontent.com/2877584/123306691-0cd4ee00-d53b-11eb-8314-7273f41c9d57.png)

![Снимок экрана 2021-06-24 в 22 24 06](https://user-images.githubusercontent.com/2877584/123306610-f29b1000-d53a-11eb-85a8-a33761a617e6.png)

Demos:
1. [Type: Module and .mjs extensions](https://vercel-9a6jjx5eu-ponomarevlad.vercel.app) ([Public logs](https://vercel-9a6jjx5eu-ponomarevlad.vercel.app/_logs))
2. [Index.js imports .mjs without Type: Module](https://vercel-50z9xp2ey-ponomarevlad.vercel.app) ([Public logs](https://vercel-50z9xp2ey-ponomarevlad.vercel.app/_logs))
3. [And Lit SSR demo with basic ESM imports](https://lit-ssr-hcmrbt7o8-ponomarevlad.vercel.app) ([Public logs](https://lit-ssr-hcmrbt7o8-ponomarevlad.vercel.app/_logs))

Demo branches:
https://github.com/PonomareVlad/Vercel/tree/canary-esm-test (1, 2)
https://github.com/PonomareVlad/VerceLitSSR/tree/vercel-canary-test (3)

**Reply by styfle** [reply]
Thanks for the repro steps!

This will be fixed in PR #6399

**Reply by styfle** [reply]
Let's try this: https://github.com/PonomareVlad/VerceLitSSR/pull/1

**Reply by PonomareVlad** [reply]
> Let's try this: [PonomareVlad/VerceLitSSR#1](https://github.com/PonomareVlad/VerceLitSSR/pull/1)

It very awesome, @styfle !

Thank you so much, for your job ! 🙂

_p.s. Sorry for late answer._

**Comment by user72356** [comment]
I still have this problem with the latest version of Vercel installed. Has this ever been pushed to prod?

**Reply by PonomareVlad** [reply]
Yep, you can show your repo to identify the problem

**Reply by user72356** [reply]
Actually I'm using Typescript and it looks like they've only fixed it for Javascript at the moment. https://github.com/vercel/vercel/pull/7954
